fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -709,4 +714,183 @@ public Void visitLogicalTableScan(OptExpression optExpression, Void context) {
+            } else if (compound.isOr()) {
+                return isAlwaysFalse(compound.getChild(0)) && isAlwaysFalse(compound.getChild(1));
+            } else if (compound.isNot()) {
+                return isAlwaysTrue(predicate.getChild(0));
+            }
+        }
+        return false;
+    }
+
+    public static boolean isAlwaysTrue(ScalarOperator predicate) {
+        if (predicate instanceof ConstantOperator) {
+            ConstantOperator constant = (ConstantOperator) predicate;
+            if (constant.getType() == Type.BOOLEAN && constant.getBoolean() == true) {
+                return true;
+            }
+        } else if (predicate instanceof CompoundPredicateOperator) {
+            CompoundPredicateOperator compound = (CompoundPredicateOperator) predicate;
+            if (compound.isAnd()) {
+                return isAlwaysTrue(compound.getChild(0)) && isAlwaysTrue(compound.getChild(1));
+            } else if (compound.isOr()) {
+                return isAlwaysTrue(compound.getChild(0)) || isAlwaysTrue(compound.getChild(1));
+            } else if (compound.isNot()) {
+                return isAlwaysFalse(predicate.getChild(0));
+            }
+        }
+        return false;
+    }
+
+    public static ScalarOperator splitOr(ScalarOperator src, ScalarOperator target) {
+        List<ScalarOperator> srcItems = Utils.extractDisjunctive(src);
+        List<ScalarOperator> targetItems = Utils.extractDisjunctive(target);
+        int srcLength = srcItems.size();
+        int targetLength = targetItems.size();
+        for (ScalarOperator item : srcItems) {
+            removeAll(targetItems, item);
+        }
+        if (targetItems.isEmpty() && srcLength == targetLength) {
+            // it is the same, so return true constant
+            return ConstantOperator.createBoolean(true);
+        } else if (!targetItems.isEmpty()) {
+            // the target has more or item, so return src
+            return src;
+        } else {
+            return null;
+        }
+    }
+
+    public static void removeAll(List<ScalarOperator> scalars, ScalarOperator predicate) {
+        Iterator<ScalarOperator> iter = scalars.iterator();
+        while (iter.hasNext()) {
+            ScalarOperator current = iter.next();
+            if (current.equals(predicate)) {
+                iter.remove();
+            }
+        }
+    }
+
+    public static boolean isAllEqualInnerJoin(OptExpression root) {
+        Operator operator = root.getOp();
+        if (!(operator instanceof LogicalOperator)) {
+            return false;
+        }
+        if (operator instanceof LogicalJoinOperator) {
+            LogicalJoinOperator joinOperator = (LogicalJoinOperator) operator;
+            boolean isEqualPredicate = isColumnEqualPredicate(joinOperator.getOnPredicate());
+            if (joinOperator.getJoinType() == JoinOperator.INNER_JOIN && isEqualPredicate) {
```

Comment:
why if only one is true, you will return the true directly?

Reply:
yes, you are right. It is a bug, I will fix it.